ton_client/client/src/tvm/execute_message.rs

```diff
@@ -0,0 +1,331 @@
+pub struct ParamsOfExecuteMessage {
+    /// Input message.
+    pub message: MessageSource,
+    /// Account BOC. Must be encoded as base64.
+    pub account: String,
+    /// Execution mode.
+    pub mode: ExecutionMode,
+    /// Execution options.
+    pub execution_options: Option<ExecutionOptions>,
+}
+
+#[derive(Serialize, Deserialize, ApiType, Debug, PartialEq, Clone)]
+pub struct ResultOfExecuteMessage {
+    /// Parsed transaction.
+    ///
+    /// In addition to the regular transaction fields there is a
+    /// `boc` field encoded with `base64` which contains source
+    /// transaction BOC.
+    pub transaction: Option<Value>,
+
+    /// List of parsed output messages.
+    ///
+    /// Similar to the `transaction` each message contains the `boc`
+    /// field.
+    pub out_messages: Vec<Value>,
+
+    /// Optional decoded message bodies according to the optional
+    /// `abi` parameter.
+    pub decoded: Option<DecodedOutput>,
+
+    /// JSON with parsed updated account state. Attention! When used in
+    /// `TvmOnly` mode only data in account state is updated.
+    pub account: Option<Value>,
+}
+
+use crate::boc::{parse_account, parse_message, parse_transaction, ParamsOfParse, ResultOfParse};
+use crate::processing::parsing::decode_output;
+use crate::tvm::check_transaction::check_transaction_status;
+use std::sync::Arc;
+use ton_block;
+use ton_block::{Message, Serializable, Transaction};
+use ton_sdk::call_tvm;
+use ton_types::{Cell, SliceData};
+
+struct ExecutionOutput {
+    transaction: Option<Transaction>,
+    messages: Option<Vec<Message>>,
+    account: Option<Cell>,
+}
+
+fn parse_object<S: Serializable>(
+    context: &Arc<ClientContext>,
+    object: &S,
+    name: &str,
+    parser: fn(Arc<ClientContext>, ParamsOfParse) -> ApiResult<ResultOfParse>,
+) -> ApiResult<Value> {
+    Ok(parser(
+        context.clone(),
+        ParamsOfParse {
+            boc: serialize_object_to_base64(object, name)?,
+        },
+    )?
+    .parsed)
+}
+
+impl ExecutionOutput {
+    fn with_transaction(transaction: Transaction, account: Cell) -> Self {
+        Self {
+            transaction: Some(transaction),
+            messages: None,
+            account: Some(account),
+        }
+    }
+
+    fn with_messages(messages: Vec<Message>, account: Option<Cell>) -> Self {
+        Self {
+            transaction: None,
+            messages: Some(messages),
+            account,
+        }
+    }
+
+    fn convert_to_result(
+        self,
+        context: &Arc<ClientContext>,
+        abi: Option<&Abi>,
+    ) -> ApiResult<ResultOfExecuteMessage> {
+        let out_messages = self.convert_out_messages(context)?;
+        let decoded = if let Some(abi) = abi {
+            Some(decode_output(context, abi, &out_messages)?)
+        } else {
+            None
+        };
+        Ok(ResultOfExecuteMessage {
+            out_messages,
+            transaction: if let Some(transaction) = &self.transaction {
```

Comment:
In think `.map(...).transpose(...)` will be better than `if let ... {} else {}`

Reply:
field conversions extracted into own methods